userscript-preferences.js

```diff
@@ -0,0 +1,230 @@
+                    let label   = pref.options[j].label || key;
+                    let checked = ( this_value.hasOwnProperty(key) && this_value[key] ) ? "checked=checked" : "";
+                    dd.append(
+                        "<label>"+
+                        `<input data-userscript-pref='${this_id}' name='${key}' type='checkbox' value=1 ${checked} />`+
+                        label+
+                        "</label>"+
+                        "<br />"
+                    );
+                }
+                break;
+            case "radio":
+                // let item_id = this_id + "_" + i;
+                for ( let j=0; j<pref.options.length; j++ ) {
+                    let value = pref.options[j];
+                    let checked = value === this_value ? "checked=checked" : "";
+                    dd.append(
+                        "<label>"+
+                        `<input data-userscript-pref='${this_id}' name='${this_id}' type='radio' value='${value}' ${checked} />`+
+                        value+
+                        "</label>"+
+                        "<br />"
+                    );
+                }
+                break;
+        }
+    }
+
+    let save = $( `<button data-userscript-save='${def.key}'>Save "${label}" Settings</button>` );
+    save.click( function() {
+        _save_userscript_settings(def);
+    } );
+
+    core_prefs.append(
+        "<dt></dt>",
+        $("<dd></dd>").append( save )
+    );
+}
+
+function _save_userscript_settings(def) {
+    "use strict";
+
+    let values = {};
+
+    for ( let i=0; i<def.options.length; i++ ) {
+        let pref    = def.options[i];
+        let this_id = def.key + "_" + pref.key;
+
+        switch ( pref.type ) {
+            case "text":
+                values[ pref.key ] = $(`input[data-userscript-pref=${this_id}]`).first().val();
+                break;
+            case "checkbox":
+                let is_checked = $(`input[data-userscript-pref=${this_id}]`).first().prop("checked");
+                values[ pref.key ] = is_checked ? true : false;
+                break;
+            case "array_checkbox":
+                let all_checkboxes = $(`input[data-userscript-pref=${this_id}]`);
+                values[ pref.key ] = {};
+                all_checkboxes.map( function() {
+                    values[ pref.key ][ $(this).prop("name") ] = $(this).prop("checked") ? true : false;
+                } );
+                break;
+            case "radio":
+                values[ pref.key ] = $(`input[data-userscript-pref=${this_id}]:checked`).val();
+                break;
+
+        }
+    }
+
+    localStorage.setItem(
+        def.key,
+        JSON.stringify( values )
+    );
+
+    if ( def.hasOwnProperty("callback") ) {
+        def["callback"]( values );
+    }
+
+    $( `button[data-userscript-save=${def.key}]` ).first().append(
+        "<span class='fa fa-check-circle' style='color: green; padding-left: 0.5em'></span>"
+    );
```

Comment:
This "progress" icon should probably be shown before it starts saving the data -- that way, if there's an issue during the save process, it's more apparent to the user (vs. just waiting for everything to complete before showing the icon for a constant 1 second).